# PA005: High Value Customer Identification ( Insiders )

# 0.0 Imports

In [1]:
import re
import s3fs
import sqlite3
import  psycopg2

import pandas                                                                             as pd
import numpy                                                                              as np
import seaborn                                                                            as sns
import umap.umap_                                                                         as umap

from sqlalchemy                    import create_engine
from datetime                      import datetime

from matplotlib                    import pyplot                                          as plt

from sklearn                       import cluster                                         as c
from sklearn                       import metrics                                         as met
from sklearn                       import decomposition                                   as dd
from sklearn                       import ensemble                                        as en
from sklearn                       import mixture                                         as mx
from plotly                        import express                                         as px
from sklearn                       import preprocessing                                   as pp


import warnings
warnings.filterwarnings("ignore")

##  0.2 Load Dataset


In [2]:
# load data
# Local
# df_raw = pd.read_csv(r'~/repositorio/insiders_clustering/data/raw/Ecommerce.csv', encoding='iso-8859-1')

# S3
path_s3 ='s3://insiders-clustering/'
df_raw = pd.read_csv(path_s3 + 'Ecommerce.csv', encoding='iso-8859-1')

# # # drop extra column
df_raw = df_raw.drop(columns=['Unnamed: 8'] , axis=1)


In [3]:
df_raw.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  InvoiceDate  UnitPrice  CustomerID         Country  
0   29-Nov-16       2.55     17850.0  United Kingdom  
1   29-Nov-16       3.39     17850.0  United Kingdom  
2   29-Nov-16       2.75     17850.0  United Kingdom  
3   29-Nov-16       3.39     17850.0  United Kingdom  
4   29-Nov-16       3.39     17850.0  United Kingdom

# <font color ='red'> 1.0 Descrição dos dados </font>

In [4]:
df1 = df_raw.copy()

## 1.1 Rename Columns

In [5]:
cols_new = ['invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date','unit_price', 'customer_id', 'country']

df1.columns = cols_new

## 1.2 Data dimensios

In [6]:
print('Number of rows: {}'.format(df1.shape[0]))
print('Number of columns: {}'.format(df1.shape[1]))

Number of rows: 541909
Number of columns: 8


## 1.3 Data types

In [7]:
df1.dtypes

invoice_no       object
stock_code       object
description      object
quantity          int64
invoice_date     object
unit_price      float64
customer_id     float64
country          object
dtype: object

## 1.4 Check NA

In [8]:
df1.isna().sum()

invoice_no           0
stock_code           0
description       1454
quantity             0
invoice_date         0
unit_price           0
customer_id     135080
country              0
dtype: int64

## 1.5 Replace NA


In [9]:
df_missing = df1.loc[df1['customer_id'].isna(), :]
df_not_missing = df1.loc[~df1['customer_id'].isna(), :]



In [10]:
# create referance 

df_backup = pd.DataFrame(df_missing['invoice_no'].drop_duplicates())
df_backup['customer_id'] = np.arange(19000 , 19000+len(df_backup), 1)
df_backup.head()

# merge original with reference dataframe
df1 = pd.merge(df1, df_backup , on='invoice_no' , how='left')

#coalesce
df1['customer_id'] = df1['customer_id_x'].combine_first(df1['customer_id_y'])

# drop extra columns

df1 = df1.drop(columns=['customer_id_x' , 'customer_id_y' ] , axis = 1)

df1.head()

invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price         country  customer_id  
0    29-Nov-16        2.55  United Kingdom      17850.0  
1    29-Nov-16        3.39  United Kingdom      17850.0  
2    29-Nov-16        2.75  United Kingdom      17850.0  
3    29-Nov-16        3.39  United Kingdom      17850.0  
4    29-Nov-16        3.39  United Kingdom      17850.0

## 1.6 Changes dtypes


In [11]:
# invoice date

df1['invoice_date'] = pd.to_datetime(df1['invoice_date'] , format='%d-%b-%y')

# customer_id

df1['customer_id'] = df1['customer_id'].astype('int64')


## 1.7 Descriptive Statistics

In [12]:
num_attributes = df1.select_dtypes(include=['int64' , 'float64'])
cat_attributes = df1.select_dtypes(exclude=['int64' , 'float64','datetime64[ns]'])


### 1.7.1 Numerical Attributs

In [13]:
# central tendency - mean, median
mean = pd.DataFrame(num_attributes.apply(np.mean)).T
median = pd.DataFrame(num_attributes.apply(np.median)).T

# dispersion - std , min , max , range , skew , kurtosis
std = pd.DataFrame(num_attributes.apply(np.std)).T
mi = pd.DataFrame(num_attributes.apply(np.min)).T
ma = pd.DataFrame(num_attributes.apply(np.max)).T
ran = pd.DataFrame(num_attributes.apply(lambda x: x.max() - x.min())).T
skew = pd.DataFrame(num_attributes.apply(lambda x: x.skew())).T
kurtosis = pd.DataFrame(num_attributes.apply(lambda x: x.kurtosis())).T

# concatenate

m = pd.concat([mi , ma ,ran, mean , median  , std,  skew ,  kurtosis]).T.reset_index()
m.columns = ['Attributes' , 'min' , 'max', 'range', 'mean', 'median', 'std', 'skew', 'kurtoses']
m

Attributes       min      max      range          mean    median  \
0     quantity -80995.00  80995.0  161990.00      9.552250      3.00   
1   unit_price -11062.06  38970.0   50032.06      4.611114      2.08   
2  customer_id  12346.00  22709.0   10363.00  16688.840453  16249.00   

           std        skew       kurtoses  
0   218.080957   -0.264076  119769.160031  
1    96.759764  186.506972   59005.719097  
2  2911.408666    0.487449      -0.804287

### 1.7.2 Categorical Attributs

#### Invoice No

In [14]:
# identificação:
df_letter_invoices = df1.loc[df1['invoice_no'].apply(lambda x: bool(re.search('[^0-9]+', x))) , :]
df_letter_invoices.head()

print('Total number of invoices: {}'.format(len(df_letter_invoices)))
print('Total number of negative quantity: {}'.format(len(df_letter_invoices['quantity'] < 0)))

Total number of invoices: 9291
Total number of negative quantity: 9291


#### Stock Code

In [15]:
# check stock code only characters
df1.loc[cat_attributes['stock_code'].apply(lambda x: bool(re.search('^[a-zA-Z]+$', x))),'stock_code'].unique()

# Ação:
## 1. Remove stock_code in ['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY','DCGSSGIRL', 'PADS', 'B', 'CRUK']


array(['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'PADS', 'B', 'CRUK'], dtype=object)

# <font color ='red'> 2.0 Filtragem de variaveis </font>

In [16]:
df2 = df1.copy()

In [17]:
# ==== Numerical Attributes ====

df2 = df2.loc[df2['unit_price'] > 0.04 , :]


# ==== Categorical Attributes ====


df2 = df2[~df2['stock_code'].isin(['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY','DCGSSGIRL', 'PADS', 'B', 'CRUK'])]

# description
df2 = df2.drop('description', axis =1)

# map
df2 = df2[~df2['country'].isin (['European Community' , 'Unspecified'])]

# bad users
df2 = df2[~df2['customer_id'].isin( [16446]) ]


## quantity 
df2_returns = df2.loc[df2['quantity'] < 0, :]
df2_purchase = df2.loc[df2['quantity'] >= 0, :]


# <font color ='red'> 3.0 Feature Engeneering </font>

In [18]:
df3 = df2.copy()

## 3.1 Feature Creation

In [19]:
# data reference

df_ref = df3.drop(['invoice_no','stock_code','quantity', 'invoice_date','unit_price','country'], axis = 1 ).drop_duplicates(ignore_index = True)


### 3.1.1. Gross Revenue

In [20]:
# Gross Revenue ( Faturamento = quantity * price)
df2_purchase.loc[: , 'gross_revenue'] = df2_purchase.loc[: , 'quantity'] * df2_purchase.loc[: , 'unit_price']

df_monetary = df2_purchase.loc[: , ['customer_id' , 'gross_revenue']].groupby('customer_id').sum().reset_index()
df_ref = pd.merge(df_ref, df_monetary,on='customer_id',how='left')

df_ref.isna().sum()

customer_id       0
gross_revenue    91
dtype: int64

### 3.1.2. Recency - Day from last purchase

In [21]:
# Recency - Last day purchase
df_recency = df2_purchase.loc[: ,[ 'customer_id', 'invoice_date']].groupby('customer_id').max().reset_index()
df_recency['recency_days'] = (df2['invoice_date'].max() - df_recency['invoice_date']).dt.days
df_recency = df_recency[['customer_id','recency_days']].copy()
df_ref = pd.merge(df_ref , df_recency , on='customer_id', how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
dtype: int64

### 3.1.3. Quantity of purchased

In [22]:
# Numero de produtos
df_freq = (df2_purchase.loc[: ,['customer_id','invoice_no']].drop_duplicates()
                                                            .groupby('customer_id')
                                                            .count()
                                                            .reset_index()
                                                            .rename(columns={'invoice_no' : 'qtde_invoices'}))

df_ref = pd.merge(df_ref , df_freq , on='customer_id', how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
dtype: int64

### 3.1.4. Quantity total of items purchased


In [23]:
df_freq = df2_purchase.loc[: ,['customer_id','quantity']].groupby('customer_id').sum().reset_index().rename(columns = {'quantity' : 'qtde_items'})
df_ref = pd.merge(df_ref , df_freq , on='customer_id', how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
dtype: int64

### 3.1.5. Quantity of products purchased


In [24]:
df_freq = df2_purchase.loc[: ,['customer_id','stock_code']].groupby('customer_id').count().reset_index().rename(columns = {'stock_code' : 'qtde_products'})
df_ref = pd.merge(df_ref , df_freq , on='customer_id', how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
dtype: int64

### 3.1.6. Avg Ticket Value

In [25]:
# Avg ticket
df_avg_ticket = df2_purchase.loc[: ,['customer_id' , 'gross_revenue']].groupby('customer_id').mean().reset_index().rename(columns={'gross_revenue':'avg_ticket'})
#df_avg_ticket['avg_ticket'] = np.round(df_avg_ticket['avg_ticket'] , 2)
df_ref = pd.merge(df_ref, df_avg_ticket,on='customer_id',how='left')
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
avg_ticket       91
dtype: int64

### 3.1.8. Frequency Purchase

In [26]:
df3_purchase = df2_purchase.copy()

In [27]:
df_aux = df3_purchase[['invoice_no','customer_id', 'invoice_date']].drop_duplicates().groupby('customer_id').agg(max_  = ('invoice_date', 'max'), 
                                                                                               min_  = ('invoice_date', 'min'), 
                                                                                               days_ = ('invoice_date', lambda x: (( x.max() - x.min() ).days)+ 1),
                                                                                               buy_  = ('invoice_date', 'count')).reset_index()

# Frequency 
df_aux['frequency'] = df_aux[['buy_', 'days_']].apply( lambda x: x['buy_'] / x['days_'] if x['days_'] != 0 else 0, axis = 1)

# Merge
df_ref = pd.merge(df_ref , df_aux[['customer_id', 'frequency']], on='customer_id', how = 'left')

df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
avg_ticket       91
frequency        91
dtype: int64

### 3.1.9. Numbers of Returns

In [28]:
df_returns = df2_returns[['customer_id' , 'quantity']].groupby('customer_id').sum().reset_index().rename( columns={'quantity' : 'qtde_returns'})
df_returns['qtde_returns'] = df_returns['qtde_returns'] * -1
df_ref = pd.merge(df_ref , df_returns, on ='customer_id' ,  how='left')

df_ref.loc[df_ref['qtde_returns'].isna() , 'qtde_returns'] = 0
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
avg_ticket       91
frequency        91
qtde_returns      0
dtype: int64

# <font color ='red'> 4.0 EDA (Exploratory Data Analysis) </font>

In [29]:
df4 = df_ref.dropna().copy()

## 4.3 Estudo do espaço

In [30]:
# Selected Dataset
cols_selected = ['customer_id', 'gross_revenue','recency_days', 'qtde_products', 'frequency', 'qtde_returns' ]
df43 = df4[cols_selected].drop(columns = 'customer_id')


In [31]:
mm = pp.MinMaxScaler()

df43['gross_revenue']          = mm.fit_transform(df43[['gross_revenue']])
df43['recency_days']           = mm.fit_transform(df43[['recency_days']])
df43['qtde_products']          = mm.fit_transform(df43[['qtde_products']])
df43['frequency']              = mm.fit_transform(df43[['frequency']])
df43['qtde_returns']           = mm.fit_transform(df43[['qtde_returns']])


X = df43.copy()


In [32]:
# training dataset 
X = df43.drop(columns=[ 'gross_revenue'], axis = 1)
Y = df43['gross_revenue']

# model definition
rf_model = en.RandomForestRegressor(n_estimators = 300 , random_state = 42 )

# model training
rf_model.fit(X,Y)

# dataframe Leaf
df_leaf = pd.DataFrame(rf_model.apply(X))

In [33]:
# tree
num_comp=[2,3,4,5]

components_list1=[]
components_list2=[]
components_list3=[]
components_list4=[]

df_tree_1 = pd.DataFrame()
df_tree_2 = pd.DataFrame()
df_tree_3 = pd.DataFrame()
df_tree_4 = pd.DataFrame()

for i in num_comp:
    
    # reduzer dimensionality
    reducer = umap.UMAP(random_state = 42, n_components = i)
    embedding = reducer.fit_transform(df_leaf)

    
    if i == 2:
        components_list1.append(embedding)
        # embedding
        df_tree_1['embedding_x'] = embedding[:,0]
        df_tree_1['embedding_y'] = embedding[:,1]


        
    if i == 3:
        components_list2.append(embedding)
        # embedding
        df_tree_2['embedding_x'] = embedding[:,0]
        df_tree_2['embedding_y'] = embedding[:,1]
        df_tree_2['embedding_z'] = embedding[:,2]

        
    if i == 4:
        components_list3.append(embedding)
        df_tree_3['embedding_x'] = embedding[:,0]
        df_tree_3['embedding_y'] = embedding[:,1]
        df_tree_3['embedding_z'] = embedding[:,2]
        df_tree_3['embedding_w'] = embedding[:,3]

    if i == 5:
        components_list4.append(embedding)
        df_tree_4['embedding_x'] = embedding[:,0]
        df_tree_4['embedding_y'] = embedding[:,1]
        df_tree_4['embedding_z'] = embedding[:,2]
        df_tree_4['embedding_w'] = embedding[:,3]
        df_tree_4['embedding_l'] = embedding[:,4]


# <font color ='red'> 5.0 Data Preparation </font>

In [34]:
# UMAP EMbedding
df5 = df_tree_4.copy()
df5.to_csv('../src/data/tree_based_embedding.csv' , index = False)

# <font color ='red'> 6.0 Hyperparameter Fine-Tunning </font>

In [35]:
x = df5.copy()

In [36]:
clusters = np.arange(2,25,1)

In [37]:
kmeans_list = []

for k in clusters:

    # model definition
    kmeans_model = c.KMeans(n_clusters = k , random_state=42)

    # model training
    kmeans_model.fit(x)

    # model predict
    labels = kmeans_model.predict(x)

    # model performance
    sil = met.silhouette_score(x , labels , metric='euclidean')

    kmeans_list.append(sil)

# <font color ='red'> 7.0 Model Training </font>

In [38]:
df8 = x.copy()

## 7.1 GMM

In [39]:
# model definition
k = 11
kmeans = c.KMeans(init='random', n_clusters=k, n_init=10, max_iter=300, random_state=42)

# model training
kmeans.fit(df8)

# clustering
labels = kmeans.labels_



## 7.2 Cluster Validation

In [40]:
print('SS value: {}'.format(met.silhouette_score(df8, labels, metric='euclidean')))

SS value: 0.518696665763855


# <font color ='red'> 8.0 Cluster Analysis </font>

In [41]:
df92 = df4[cols_selected].copy()
df92['cluster'] = labels



df92['last_training_timestamp'] = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

# change dtypes
df92['recency_days'] = df92['recency_days'].astype('int64')
df92['qtde_products'] = df92['qtde_products'].astype('int64')
df92['qtde_returns'] = df92['qtde_returns'].astype('int64')
df92['last_training_timestamp'] =pd.to_datetime(df92['last_training_timestamp'])

## 7.2 Cluster Profile


In [42]:
# Number of customer
df_cluster = df92[['customer_id','cluster']].groupby('cluster').count().reset_index()
df_cluster['perc_customers'] = 100*(df_cluster['customer_id'] / df_cluster['customer_id'].sum())

# Avg Gross revenue
df_avg_gross_revenue = df92[['gross_revenue','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge(df_cluster , df_avg_gross_revenue , how='inner' , on='cluster')

# Avg recency days
df_avg_recency_days = df92[['recency_days','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge(df_cluster , df_avg_recency_days , how='inner' , on='cluster')

# Avg invoice_no
df_qtde_products = df92[['qtde_products','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge(df_cluster , df_qtde_products , how='inner' , on='cluster')

# Frequency
df_frequency = df92[['frequency','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge(df_cluster , df_frequency , how='inner' , on='cluster')

# Returns
df_qtde_returns = df92[['qtde_returns','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge(df_cluster , df_qtde_returns , how='inner' , on='cluster')


df_cluster

cluster  customer_id  perc_customers  gross_revenue  recency_days  \
0         0          548        9.622476    1122.025712     65.476277   
1         1          807       14.170325    2174.007162     94.013631   
2         2          471        8.270413     145.347197    154.131635   
3         3          580       10.184372     177.296345    192.812069   
4         4          400        7.023705     724.093875    111.145000   
5         5          279        4.899034    1364.278280     83.455197   
6         6          691       12.133450    7350.111245     54.269175   
7         7          267        4.688323     416.195019    165.172285   
8         8          437        7.673398    1674.142929     85.828375   
9         9          599       10.517998     304.986277    166.297162   
10       10          616       10.816506    1281.040097    135.173701   

    qtde_products  frequency  qtde_returns  
0       45.233577   0.026652      7.759124  
1      161.778191   0.428010      0.591078  
2        4.106157   0.917057      3.855626  
3        7.343103   1.002972      4.987931  
4       32.995000   0.454489      5.880000  
5       66.971326   0.119078     14.799283  
6      375.098408   0.253673    222.586107  
7       30.902622   1.012588      0.816479  
8       30.244851   0.035986      9.116705  
9       17.333890   1.014218      2.425710  
10      69.613636   0.745552      4.370130

In [52]:
# 6 cluster Insiders
# 1 cluster more products
# 8 cluster spend money
# 5 cluster even more products
# 10 cluster less days
# 0 cluster less 1k
# 4 cluster Stop Returners
# 7 cluster more buy
# 9 cluster even more buy
# 3 cluster weak
# 2 cluster latecomer

### Cluster 01: ( Candidato a Insiders)

    - Número de customers: 6915 (12% do customers)
    - Faturamento médio: 7350
    - Recência média: 54 dias 
    - Média de produtos comprados: 375 produtos
    - Frequência de produtos comprados: 0.25 produtos/dia
    - Receita em média: $ 7350,11 dólares
    


# <font color ='red'> 8.0 Deploy to production </font>

In [43]:
df92.dtypes

customer_id                         int64
gross_revenue                     float64
recency_days                        int64
qtde_products                       int64
frequency                         float64
qtde_returns                        int64
cluster                             int32
last_training_timestamp    datetime64[ns]
dtype: object

## 8.1. Insert into SQLITE

In [46]:
# database connection
# endpoint = 'sqlite:///insiders_db.sqlite'
host = 'database-insiders.csgyb6dz6l8q.us-east-1.rds.amazonaws.com'
port = '5432'
database = 'postgres'
user = 'fimasterdb'
pwd = 'PeD41786325829'

# endpoint = 'postgresql://user:pwd@host:port'
endpoint = 'postgresql://fimasterdb:PeD41786325829@database-insiders.csgyb6dz6l8q.us-east-1.rds.amazonaws.com:5432/postgres'
conn = create_engine(endpoint)

# drop table
# query_drop_insiders="""
#     DROP TABLE Insiders
# """


# create Table
# query_create_table_insiders = """
#     CREATE TABLE Insiders (
#         customer_id                   INTEGER,
#         gross_revenue                 REAL,
#         recency_days                  INTEGER,
#         qtde_products                 INTEGER,
#         frequency                     REAL,
#         qtde_returns                  INTEGER,
#         cluster                       INTEGER,
#         last_training_timestamp       TIMESTAMP
#      )

# """

# conn.execute(query_create_table_insiders)

In [47]:
# insert data into
df92.to_sql('Insiders', con=conn, if_exists='append' , index=False)



AttributeError: 'Engine' object has no attribute 'close'

In [ ]:
# get query
# query_collect="""
#     SELECT * from Insiders
# """

# df = pd.read_sql_query(query_collect , conn)
# df.head()